Aggregate journals in APIv3

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/20688
## Description

Starting with this PR the APIv3 activities return their results in an aggregated form.
Some work on the `Journal::AggregatedJournals` was still neccessary.

**WP legcy form (non-aggregated)**
![non-aggregated](https://cloud.githubusercontent.com/assets/453584/8877643/a1349582-3226-11e5-853b-9e7bc6afe8e3.PNG)

**inplace view (aggregated)**
![aggregated](https://cloud.githubusercontent.com/assets/453584/8877642/a1343b32-3226-11e5-9a44-7a4558b6253a.PNG)

NobodysNightmare committed:
extract get_changes into own module

NobodysNightmare committed:
migrate APIv3 to aggregated journals

- turns out agrgegated journals needed more features
- turns out specs were incomplete and bogus

NobodysNightmare committed:
fix predecessor implementation

NobodysNightmare committed:
fix update link

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-aggregated-api-journals.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-aggregated-api-journals.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

oliverguenther commented:
I've restarted the specs thrice in the course over late Friday / Saturday as they interfered with my spec run. The Postgres timeout is persistent. Some change seems to successfully lock up the DB.

NobodysNightmare committed:
improve exception message on missing journable.id

previously an invalid MySQL statement would have been indicated,
which is correct but not very helpful.
On PSQL this will also lead to a DB lockup (at least on environments known to me)

NobodysNightmare commented:
PostgreSQL is weird... I already encountered this earlier: When you have multiple specs with a bad SQL syntax it will (after ~5 bad queries) start to refuse further connections, letting rails time out on each and every test case.

For travis this means hitting the 120 minute mark, never telling anything about the connection timeouts.

oliverguenther commented:
Travis did tell me about a timeout on Postgres, something like `couldn't get stable pg connection for 5 minutes ..`, though these messages seem to disappear.

NobodysNightmare committed:
add id to built models

tried to correctly "build_stubbed" them, but turns out that this is harder than I would like it to be...

NobodysNightmare committed:
fix C&P error

NobodysNightmare committed:
extract activities by WP into own endpoint

Azure7111 commented on app/models/journal/aggregated_journal.rb:
```diff
@@ -182,21 +202,31 @@ def table_name
-                    .where("#{Journal.table_name}.version < ?", version)
-                    .order("#{Journal.table_name}.version DESC")
+                    .where("#{version_sql} < ?", version)
+                    .except(:order)
```
Do I get this right: This line removes, or rather prevents, an ordering which happens inside `query_aggregated_journals`?

NobodysNightmare replied:
Right. I had to look this up myself:

When you `order`, then AR will only **append** that order to the existing order:
`order(id: :asc).order(name: :desc) == 'ORDER BY id ASC, name DESC'`

By using `except(:order)` you throw away anything there was about the order and start from scratch.

The old (deprecated?) way of doing it is `reorder`

Azure7111 replied:
Given the fact that orders are _not_ "overwritten" this makes actually sense to me now. Thanks!

Azure7111 commented on lib/api/v3/activities/activities_by_work_package_api.rb:
```diff
@@ -0,0 +1,65 @@
+              if work_package.save
+                journals = ::Journal::AggregatedJournal.aggregated_journals(
+                  journable: work_package)
+                Activities::ActivityRepresenter.new(journals.last, current_user: current_user)
```
Can we be sure that no other journals have been created since the last line?

NobodysNightmare replied:
I don't think so. But:
- I think there is no quick way to do it better
- I did not introduce this problem, it was there before

This does not mean it shouldn't be solved, it just does not belong here and I hope that it is for now unimportant enough.

Azure7111 commented on lib/api/v3/activities/activity_representer.rb:
```diff
@@ -47,20 +48,20 @@ class ActivityRepresenter < ::API::Decorators::Single
           {
-            href: api_v3_paths.activity(represented.id),
-            method: :patch,
-            title: "#{represented.id}"
```
Why did you remove the link titles?

NobodysNightmare replied:
Because `16435` is not a title I would present to a user.

And I wont introduce a proper shiny formatting (including localization) for a title that no one (assumption) will ever display.

Azure7111 commented on lib/plugins/acts_as_journalized/lib/journal_changes.rb:
```diff
@@ -0,0 +1,78 @@
+#-- encoding: UTF-8
+module JournalChanges
+  def get_changes
+    return {} if data.nil?
```
I'm not sure about this: Is there any way to ensure that the including class of this module has a `data` method?

NobodysNightmare replied:
In case it does not have one, a `NoMethodError` will be thrown to indicate that you did it wrong xD

I guess the answer here is "it is a dynamic world, deal with it".
I've seen annotations and stuff in other environments, but that requires your tools to support you.

The other option one could choose is to pass all needed values as parameter into `get_changes`. However, this is an approach I would use when designing a class that the caller will instantiate and hold as member (composition). In this case I chose to use a module that another class can include (which is more Trait-like). In that case I prefer to go the "expect what you need to be there" route.

Azure7111 commented on lib/plugins/acts_as_journalized/lib/journal_changes.rb:
```diff
@@ -0,0 +1,78 @@
+
+    if predecessor.nil?
+      @changes = data.journaled_attributes.select { |_, v| !v.nil? }
+                   .inject({}) { |h, (k, v)| h[k] = [nil, v]; h }
```
I have to agree, this needs more style. Please try again.

NobodysNightmare replied:
Not my darn code, I solely moved it.

Azure7111 commented on spec/requests/api/v3/activity_resource_spec.rb:
```diff
@@ -80,9 +84,40 @@
+
+  describe '#patch' do
+    subject(:response) { last_response }
+    let(:patch_path) { api_v3_paths.activity activity.id }
```
Do you think the patch_path and the get_path are likely to diverge in the near future? Because if not, then you should be able to move it into the initialization section in the beginning of this file. But if you say that this is about consistency across the recource specs I am fine with that too.

NobodysNightmare replied:
I actually thought that it is how all specs do it and that is why I sticked to it.

Personally I like the generic `path` more. You won't issue two different requests in one spec anyway.

Azure7111 commented on spec/requests/api/v3/support/authorization.rb:
```diff
@@ -28,27 +28,16 @@
-      it 'should respond with correct type' do
-        expect(subject.body).to include_json(type.to_json).at_path('_type')
-        expect(subject.body).to be_json_eql(id.to_json).at_path('id')
-      end
```
Why have you removed these tests?

NobodysNightmare replied:
Because they are bogus.
They are not connected to the anonymous user in any way, but to "a user with sufficient permissions", whcih is always tested in the main spec.

What is not in here (and usually in the main spec) is the "user with insufficient permissions". Note that I don't use the word "anonymous user" here, because it just has nothing to do with being anonymous.

However, there is one setting that is related to being anonymous, regardless of other permissions: Receiving a 401 as anonymous user, when the setting `login_required?` is enabled. That is why I left the one remaining spec alive, because I could see its value.

The other specs only duplicate the stuff of the main specs while confusing people by using the term "anonymous user".

NobodysNightmare committed:
improve readability of JournalChanges

NobodysNightmare committed:
try to improve readability of JournalManager

- using useful names
- using rubys awesome syntax to avoid nested list accesses (e.g. k[0][1])

Azure7111 commented:
The requirements talk about "the angular client (inplace edit) is adapted accordingly". I guess that means being able to see aggregated journals in the details pane, or is there more to it?

Otherwise this LGTM :+1:

NobodysNightmare committed:
explicitly define order of groups in PSQL

Always remember kids: SQL databases do not guarantee an order if you don't ask for one!

That is why PostgreSQL sometimes found a predecessor for an aggregated journal that was not supposed to have one: Based on internal decisions the order of rough groups was not ascending (as expected implicitly) and master and addition would switch places.

ALWAYS SPECIFY AN ORDER IF YOU EXPECT ONE!

NobodysNightmare commented:
> I guess that means being able to see aggregated journals in the details pane

Right, but I did not want to write "the angular client is not broken afterwards", which is what I basically wanted to say ^^

Azure7111 commented:
Alright, then I will merge this once it's green.
